src/classes/CW_AudienceCSVUploadControllerTest.cls

```diff
@@ -0,0 +1,397 @@
+		CW_AudienceCSVUploadController uc = new CW_AudienceCSVUploadController();
+		uc.selectedSchId = sch.Id;
+		//try to upload the file with blank fields
+		try{
+			uc.csvFile = Blob.valueOf(faultyCSVString);
+			uc.filename='FaultyFile.csv';
+			List<List<string>> csvLines = CSVReader.readCSVFile(faultyCSVString);
+		}
+		catch(Exception e)
+        {
+        	exceptionHappend = true;
+        }
+		Test.stopTest();
+		
+		// none of the groups or members gets uploaded if file contains null values
+		system.assertEquals(2,[select count() from CW_Audience_Group__c where Campaign__c = :camp.Id]);
+		// assert no empty name group
+		system.assertEquals(0,[select count() from CW_Audience_Group__c where (not Name like 'Group%')]);
+		// assert skip of duplicate group
+		system.assertEquals(0,[select count() from CW_Audience_Group__c 
+								where Campaign__c = :camp.Id and Name = :group2]);
+		
+	}
+	
+	static testMethod void parseFaultyMembers()
+	{
+		CW_Campaign__c camp = TestDataUtil.createCampaign();
+		insert camp;
+		CW_Schedule__c sch = TestDataUtil.createSchedule(camp);
+		insert sch;
+		
+		List<Account> accs = new List<Account>();
+		for (Integer i=0;i<5;i++)
+		{
+			Account acc = TestDataUtil.createPersonAccount();
+			acc.Account_Name__pc = 'TestAccountName' + i;
+			accs.add(acc);
+		}
+		insert accs;
+		
+		
+		CW_Audience_Group__c audGr1 = TestDataUtil.createAudienceGroup(camp.Id);
+		audGr1.Name = group1;
+		audGr1.Is_Control_Group__c = false;
+		insert audGr1;
+		
+		CW_Audience_Group_Schedule__c ags = [SELECT Id, Audience_Group__c 
+											 FROM CW_Audience_Group_Schedule__c 
+											 WHERE Audience_Group__c=: audGr1.Id
+											 AND Schedule__c =: sch.Id];
+		
+		String memebrsCSVString = 'Group Name,Account Name' + CSVReader.lineTerminator;
+		
+		// create existing member for account 1
+		insert new CW_Audience_Group_Member__c(Audience_Group_LU__c = audGr1.Id, 
+					Contact__r = new Contact(Account_Name__c = accs[0].Account_Name__pc),
+					Audience_Group_Schedule__c = ags.Id);
+					
+		// add duplicate to CSV for account 2
+		memebrsCSVString += audGr1.Name + ',' + accs[1].Account_Name__pc + CSVReader.lineTerminator;
+		
+		Test.setCurrentPage(Page.CW_AudienceCSVUpload);
+		Apexpages.currentPage().getParameters().put(ConstantUtil.PAGEPARAM_CAMPAIGNID,camp.Id);
+		
+		Test.startTest();
+		CW_AudienceCSVUploadController uc = new CW_AudienceCSVUploadController();
+		uc.selectedSchId = sch.Id;
+		uc.csvFile = Blob.valueOf(memebrsCSVString);
+		uc.filename='Somefile.csv';
+		uc.uploadCSV();
+		uc.loadBatchJobProgressPage();
+		Test.stopTest();
+		
+		system.assertEquals(1,[select count() from CW_Audience_Group_Member__c where Audience_Group_LU__c = :audGr1.Id]);
+		// assert there are no duplicates
+		system.assertEquals(1,[select count() from CW_Audience_Group_Member__c 
+			where Contact__r.Account_Name__c = :accs.get(0).Account_Name__pc]);
+	}
+	
+	static testMethod void parseLFLineTerminator()
```

Comment:
Style issue: test methods should be annotated with @IsTest for better readibility. Style issue: methods should use the "lowerCamelCase" notation. Suggestion: rename accordingly.